test/test_autograd.py

```diff
@@ -4014,6 +4015,86 @@ def backward(ctx, grad):
         foo = MyFn.apply(base, True)
         self.assertEqual(foo.grad_fn.__class__.__name__, "MyFnBackward")
 
+    def test_engine_shared_graph(self):
+        model_shared = nn.Sequential()
+        for i in range(10):
+            model_shared.add_module(str(i), nn.Linear(5, 5))
+            model_shared.add_module(str(i) + "_relu", nn.ReLU())
+
+        model_end = nn.Linear(5, 5)
+
+        # Regular backward
+        inp = torch.rand(2, 5)
+        middle = model_shared(inp)
+        out = model_end(middle)
+        out.sum().backward()
+        saved_grads = [p.grad.clone() for p in model_shared.parameters()]
+
+        # Backward with hook that calls backward on a subset of the original graph
+        # Note that this is not a real algorithm but just an example
+        model_shared.zero_grad()
+        inp = torch.rand(2, 5)
+        middle = model_shared(inp)
+        def my_hook(grad):
+            # Run the backward and compute the 2-norm of gradients
+            grads = torch.autograd.grad(middle, model_shared.parameters(), grad, retain_graph=True)
+            norm = sum(g.norm() for g in grads)
+            # Re-scale the gradients accordingly
+            return grad / norm
+        out = model_end(middle)
+        out.register_hook(my_hook)
+        out.sum().backward()
+
+    def test_engine_shared_graph_threading(self):
+        # This test tries a "worst case scenario" to make sure that we don't call a single Node concurrently
+        # from multiple threads while multiple backward passes are running.
+        class MyFunction(torch.autograd.Function):
+            in_use = False
+
+            @staticmethod
+            def forward(ctx, inp):
+                return inp.clone()
+
+            @staticmethod
+            def backward(ctx, grad_output):
+                # Make sure nothing else enters here before we're done
+                if MyFunction.in_use:
+                    raise RuntimeError("A single node is entered concurrently")
+                MyFunction.in_use = True
+                # Note that time.sleep release the GIL
+                time.sleep(0.2)
+                MyFunction.in_use = False
+                return grad_output
+
+        class MyMod(nn.Module):
+            def forward(self, inp):
+                return MyFunction.apply(inp)
+
+        model_shared = nn.Sequential()
+        for i in range(10):
+            if i == 5:
+                model_shared.add_module(str(i) + "_custom", MyMod())
+            model_shared.add_module(str(i), nn.Linear(5, 5))
+            model_shared.add_module(str(i) + "_relu", nn.ReLU())
+
+        model_end = nn.Linear(5, 1)
+
+        inp = torch.rand(2, 5)
+        middle = model_shared(inp)
+
+        def run_backward():
+            out = model_end(middle)
+            grads = torch.autograd.grad(out.sum(), model_shared.parameters(), retain_graph=True)
```

Comment:
The expected behavior of this is that in the current master engine with no parallelism on CPU, this will never raise, but with parallel autograd engine this should raise right? Because `time.sleep(0.2)` release the GIL (we rely on GIL for PyNode to ensure thread safety rather than locking ourselves), and other tasks can enter the same Node, which will raise exception right?

Reply:
Yes, removing this to avoid merge issues with the new engine.